0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+use a similar approach as an internal optimization when many hashing collisions occur, 
+so that the worst case runtime of lookups is O(N log N) rather than O(N).
+The included Dafny runtimes can provide an implementation of Dafny sets and multisets that use this approach consistently,
+and hence provide the efficient average-case access of a hash set but still provide a deterministic enumeration order.
+
+The good news is that individual compilers are free to decide they do not support sequence comprehensions
+whose ordering depends on any such collections. Dafny code that will be compiled using such compilers can 
+fall back to using a `foreach` loop to iterate over sets instead, as statements are permitted to
+observe non-determinism. This means a particular compiler can still map a Dafny `set<T>` value 
+directly to a native implementation of sets, such as a Java `Set<T>`, even though these implementations will
+enumerate their contents with non-deterministic ordering.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The primary drawback to this proposal is the burden of adding the new concept of enumeration ordering. 
+This means extra cognitive load for Dafny users, extra implementation effort for any new builtin collection types, 
+and additional testing to ensure that all supported quantification domains are deterministically enumerable.
+
+It can be argued that if any compiled Dafny program uses existing features that depend on domain enumeration,
+such as `exists` expressions or `:|` statements, its runtime behavior already depends heavily on how this domain
+searching is implemented. The compiler only enforces that such domains are finite and enumerable, but an inefficient
+choice can lead to the search for a matching value taking orders of magnitude more time than expected.
+Therefore we should already be documenting this aspect of Dafny semantics and implementation better anyway.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The most obvious alternative is to only provide the simpler `foreach x in C` syntax, where `C` is any collection value.
+This is certainly a well-formed and sound construct, but far less powerful than the proposed version, especially
+when sequence comprehensions are excluded. It also still runs into some of the same semantic challenges around
+ordering if `C` is allowed to be a set or multiset anyway, since verification of such a loop still needs to be aware of
+this ordering. Offering `foreach` loops along with sequence comprehensions means we can recommend the latter where possible,
+with the former as a fallback that is inevitably more complicated to verify.
+
+# Prior art
+[prior-art]: #prior-art
+
+Sequence comprehensions bear a strong resemblance to list comprehensions in Python, which also include similar abilities to 
+bind multiple variables and filter the enumerated values: `[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]`
```

Comment:
I find there's a big difference to those comprehensions and the ones introduced in this RFC, namely that in Python an explicit value generator has to be specified, while in this RFC the syntax seems to specify an arbitrary boolean expression from which, using some magic, a value generator is extracted.

Reply:
Definitely. It's really only related prior art in that Python's comprehensions are values expressed in a similar way rather than loops. Sequence comprehensions are definitely more declarative and less explicit about the implementation.